File: python/src/debiased_inference/bandwidth.py
"""Bandwidth selectors for ordinary KDE and local-linear regression."""

from __future__ import annotations

import numpy as np
from numpy.typing import ArrayLike

from ._validation import as_samples, as_vector, positive_scalar


def density_bandwidth(
    x: ArrayLike,
    *,
    method: str = "normal_reference",
    candidates: ArrayLike | None = None,
    block_size: int = 512,
) -> float:
    """Select an isotropic bandwidth for the ordinary KDE.

    ``method="normal_reference"`` uses the Gaussian normal-scale rule used by
    the paper's ``ks`` implementation. In one dimension this is
    ``(4 / (3 n))^(1/5) * sd``. In higher dimensions the scalar scale is the
    geometric mean of positive marginal standard deviations and the
    normal-reference dimension adjustment is used.

    ``method="cv"`` minimizes the least-squares cross-validation criterion
    over ``candidates``. If candidates are omitted, a geometric grid centered
    on the normal-reference choice is used. Pairwise calculations are blocked
    to use ``O(n * block_size)`` rather than ``O(n^2)`` memory.
    """
    samples = as_samples(x)
    if method not in {"normal_reference", "cv"}:
        raise ValueError("method must be 'normal_reference' or 'cv'")
    n, dimension = samples.shape
    standard_deviation = np.std(samples, axis=0, ddof=1)
    positive = standard_deviation[standard_deviation > 0]
    if positive.size == 0:
        raise ValueError("cannot select bandwidth from zero-scale data")
    scale = float(np.exp(np.mean(np.log(positive))))
    factor = (4.0 / (dimension + 2.0)) ** (1.0 / (dimension + 4.0))
    factor *= n ** (-1.0 / (dimension + 4.0))
    reference = positive_scalar(scale * factor, name="selected bandwidth")
    if method == "normal_reference":
        if candidates is not None:
            raise ValueError("candidates are only used when method='cv'")
        return reference

    if candidates is None:
        candidate_values = reference * np.geomspace(0.35, 2.5, 31)
    else:
        candidate_values = as_vector(candidates, name="candidates")
        if np.any(candidate_values <= 0):
            raise ValueError("candidates must be positive")
    if not isinstance(block_size, (int, np.integer)) or block_size < 1:
        raise ValueError("block_size must be a positive integer")
    scores = np.empty(candidate_values.size, dtype=float)
    for index, candidate in enumerate(candidate_values):
        h = float(candidate)
        integrated_normalizer = (4.0 * np.pi * h**2) ** (-0.5 * dimension)
        ordinary_normalizer = (2.0 * np.pi * h**2) ** (-0.5 * dimension)
        integrated_sum = 0.0
        ordinary_sum = 0.0
        for start in range(0, n, int(block_size)):
            stop = min(start + int(block_size), n)
            differences = samples[start:stop, None, :] - samples[None, :, :]
            distances_squared = np.sum(differences**2, axis=-1)
            integrated_sum += integrated_normalizer * float(
                np.sum(np.exp(-distances_squared / (4.0 * h**2)))
            )
            ordinary_sum += ordinary_normalizer * float(
                np.sum(np.exp(-distances_squared / (2.0 * h**2)))
            )
        ordinary_off_diagonal = ordinary_sum - n * ordinary_normalizer
        scores[index] = (
            integrated_sum / n**2
            - 2.0 * ordinary_off_diagonal / (n * (n - 1))
        )
    return float(candidate_values[int(np.argmin(scores))])


def regression_bandwidth(
    x: ArrayLike,
    y: ArrayLike,
    *,
    candidates: ArrayLike | None = None,
    n_folds: int = 5,
    random_state: int | None = 0,
) -> float:
    """Select a local-linear bandwidth by deterministic K-fold CV.

    Candidate loss is mean squared prediction error from the ordinary
    local-linear smoother. Failed boundary fits are excluded; a candidate is
    eligible only if every held-out point can be predicted.
    """
    from .regression import _local_polynomial

    x_values = as_vector(x, name="x")
    y_values = as_vector(y, name="y")
    if x_values.size != y_values.size:
        raise ValueError("x and y must have the same length")
    if np.ptp(x_values) == 0:
        raise ValueError("x must contain at least two distinct values")
    n = x_values.size
    if not isinstance(n_folds, (int, np.integer)) or not 2 <= n_folds <= n:
        raise ValueError("n_folds must be an integer between 2 and len(x)")

    if candidates is None:
        scale = min(float(np.std(x_values, ddof=1)), float(np.ptp(x_values)) / 4.0)
        base = max(scale * n ** (-1.0 / 5.0), np.finfo(float).eps)
        candidate_values = base * np.geomspace(0.35, 2.5, 21)
    else:
        candidate_values = as_vector(candidates, name="candidates")
        if np.any(candidate_values <= 0):
            raise ValueError("candidates must be positive")

    generator = np.random.default_rng(random_state)
    order = generator.permutation(n)
    folds = np.array_split(order, n_folds)
    losses = np.full(candidate_values.size, np.inf)
    for candidate_index, candidate in enumerate(candidate_values):
        squared_errors: list[np.ndarray] = []
        valid = True
        for held_out in folds:
            keep = np.ones(n, dtype=bool)
            keep[held_out] = False
            predicted = _local_polynomial(
                x_values[keep], y_values[keep], x_values[held_out],
                float(candidate), degree=1, derivative=0,
            )
            if np.any(~np.isfinite(predicted)):
                valid = False
                break
            squared_errors.append((y_values[held_out] - predicted) ** 2)
        if valid:
            losses[candidate_index] = float(np.mean(np.concatenate(squared_errors)))
    if not np.any(np.isfinite(losses)):
        raise RuntimeError("all candidate bandwidths produced singular local fits")
    return float(candidate_values[int(np.argmin(losses))])

File: r/R/bandwidth.R
#' Select an ordinary density-estimation bandwidth
#'
#' Uses the Gaussian normal-scale rule used by the paper's `ks`
#' implementation. In one dimension this is `(4 / (3 n))^(1/5) * sd`.
#' Bandwidth selection is for the ordinary estimator, as required by the paper.
#' @param x Numeric observations or a numeric matrix with observations in rows.
#' @param method Either `"normal_reference"` or `"cv"`.
#' @param candidates Optional positive candidate bandwidths for cross-validation.
#' @param block_size Pairwise-computation block size for density cross-validation.
#' @return A positive numeric scalar.
#' @export
density_bandwidth <- function(x, method = "normal_reference", candidates = NULL,
                              block_size = 512L) {
  samples <- .as_samples(x)
  if (length(method) != 1L || is.na(method) ||
      !method %in% c("normal_reference", "cv")) {
    stop("method must be 'normal_reference' or 'cv'", call. = FALSE)
  }
  n <- nrow(samples)
  dimension <- ncol(samples)
  standard_deviation <- apply(samples, 2L, stats::sd)
  positive <- standard_deviation[standard_deviation > 0]
  if (length(positive) == 0L) {
    stop("cannot select bandwidth from zero-scale data", call. = FALSE)
  }
  scale <- exp(mean(log(positive)))
  factor <- (4 / (dimension + 2))^(1 / (dimension + 4)) *
    n^(-1 / (dimension + 4))
  reference <- .positive_scalar(scale * factor, "selected bandwidth")
  if (method == "normal_reference") {
    if (!is.null(candidates)) {
      stop("candidates are only used when method='cv'", call. = FALSE)
    }
    return(reference)
  }
  if (is.null(candidates)) {
    candidates <- reference * exp(seq(log(0.35), log(2.5), length.out = 31L))
  } else {
    candidates <- .as_vector(candidates, "candidates")
    if (any(candidates <= 0)) stop("candidates must be positive", call. = FALSE)
  }
  if (!is.numeric(block_size) || length(block_size) != 1L ||
      !is.finite(block_size) || block_size < 1 ||
      block_size != as.integer(block_size)) {
    stop("block_size must be a positive integer", call. = FALSE)
  }
  block_size <- as.integer(block_size)
  scores <- numeric(length(candidates))
  for (index in seq_along(candidates)) {
    h <- candidates[index]
    integrated_normalizer <- (4 * pi * h^2)^(-dimension / 2)
    ordinary_normalizer <- (2 * pi * h^2)^(-dimension / 2)
    integrated_sum <- 0
    ordinary_sum <- 0
    starts <- seq.int(1L, n, by = block_size)
    for (start in starts) {
      stop_index <- min(start + block_size - 1L, n)
      block <- samples[start:stop_index, , drop = FALSE]
      norm_squared <- matrix(0, nrow = nrow(block), ncol = n)
      for (coordinate in seq_len(dimension)) {
        differences <- outer(block[, coordinate], samples[, coordinate], "-")
        norm_squared <- norm_squared + differences^2
      }
      integrated_sum <- integrated_sum + integrated_normalizer *
        sum(exp(-norm_squared / (4 * h^2)))
      ordinary_sum <- ordinary_sum + ordinary_normalizer *
        sum(exp(-norm_squared / (2 * h^2)))
    }
    ordinary_off_diagonal <- ordinary_sum - n * ordinary_normalizer
    scores[index] <- integrated_sum / n^2 -
      2 * ordinary_off_diagonal / (n * (n - 1))
  }
  as.double(candidates[which.min(scores)])
}

#' Select a local-linear regression bandwidth
#'
#' Performs K-fold cross-validation on the ordinary local-linear estimator.
#' @param x Numeric covariates.
#' @param y Numeric responses.
#' @param candidates Optional positive candidate bandwidths.
#' @param n_folds Number of folds.
#' @param random_state Optional local random seed.
#' @return The candidate with smallest mean squared validation error.
#' @export
regression_bandwidth <- function(x, y, candidates = NULL, n_folds = 5L,
                                 random_state = 0L) {
  x <- .as_vector(x, "x")
  y <- .as_vector(y, "y")
  if (length(x) != length(y)) stop("x and y must have the same length", call. = FALSE)
  if (diff(range(x)) == 0) stop("x must contain at least two distinct values", call. = FALSE)
  n <- length(x)
  if (!is.numeric(n_folds) || length(n_folds) != 1L ||
      n_folds < 2 || n_folds > n || n_folds != as.integer(n_folds)) {
    stop("n_folds must be an integer between 2 and length(x)", call. = FALSE)
  }
  n_folds <- as.integer(n_folds)
  if (is.null(candidates)) {
    scale <- min(stats::sd(x), diff(range(x)) / 4)
    base <- max(scale * n^(-1 / 5), .Machine$double.eps)
    # Include the small CV bandwidths selected for the paper's high-curvature
    # regression designs; a narrower range can truncate the optimum.
    candidates <- base * exp(seq(log(0.1), log(3), length.out = 31L))
  } else {
    candidates <- .as_vector(candidates, "candidates")
    if (any(candidates <= 0)) stop("candidates must be positive", call. = FALSE)
  }
  fold_id <- .with_seed(random_state, {
    order <- sample.int(n)
    result <- integer(n)
    result[order] <- rep(seq_len(n_folds), length.out = n)
    result
  })
  losses <- rep(Inf, length(candidates))
  for (candidate_index in seq_along(candidates)) {
    errors <- numeric()
    valid <- TRUE
    for (fold in seq_len(n_folds)) {
      held_out <- fold_id == fold
      predicted <- .local_polynomial(
        x[!held_out], y[!held_out], x[held_out], candidates[candidate_index],
        degree = 1L, derivative = 0L
      )
      if (any(!is.finite(predicted))) {
        valid <- FALSE
        break
      }
      errors <- c(errors, (y[held_out] - predicted)^2)
    }
    if (valid) losses[candidate_index] <- mean(errors)
  }
  if (!any(is.finite(losses))) {
    stop("all candidate bandwidths produced singular local fits", call. = FALSE)
  }
  as.double(candidates[which.min(losses)])
}
